osu.Game.Rulesets.Taiko/UI/HitExplosion.cs

```diff
@@ -16,39 +18,80 @@ namespace osu.Game.Rulesets.Taiko.UI
     /// <summary>
     /// A circle explodes from the hit target to indicate a hitobject has been hit.
     /// </summary>
-    internal class HitExplosion : CircularContainer
+    internal class HitExplosion : PoolableDrawable
     {
         public override bool RemoveWhenNotAlive => true;
-
-        [Cached(typeof(DrawableHitObject))]
-        public readonly DrawableHitObject JudgedObject;
+        public override bool RemoveCompletedTransforms => false;
 
         private readonly HitResult result;
 
-        private SkinnableDrawable skinnable;
+        [CanBeNull]
+        public DrawableHitObject JudgedObject;
 
-        public override double LifetimeStart => skinnable.Drawable.LifetimeStart;
+        private SkinnableDrawable skinnable;
 
-        public override double LifetimeEnd => skinnable.Drawable.LifetimeEnd;
+        /// <summary>
+        /// This constructor only exists to meet the <c>new()</c> type constraint of <see cref="DrawablePool{T}"/>.
+        /// </summary>
+        public HitExplosion()
+            : this(HitResult.Great)
+        {
+        }
 
-        public HitExplosion(DrawableHitObject judgedObject, HitResult result)
+        public HitExplosion(HitResult result)
         {
-            JudgedObject = judgedObject;
             this.result = result;
 
             Anchor = Anchor.Centre;
             Origin = Anchor.Centre;
 
-            RelativeSizeAxes = Axes.Both;
             Size = new Vector2(TaikoHitObject.DEFAULT_SIZE);
+            RelativeSizeAxes = Axes.Both;
 
             RelativePositionAxes = Axes.Both;
         }
 
         [BackgroundDependencyLoader]
         private void load()
         {
-            Child = skinnable = new SkinnableDrawable(new TaikoSkinComponent(getComponentName(result)), _ => new DefaultHitExplosion(JudgedObject, result));
+            InternalChild = skinnable = new SkinnableDrawable(new TaikoSkinComponent(getComponentName(result)), _ => new DefaultHitExplosion(result));
+            skinnable.OnSkinChanged += runAnimation;
+        }
+
+        public void Apply([CanBeNull] DrawableHitObject drawableHitObject)
+        {
+            JudgedObject = drawableHitObject;
+        }
+
+        protected override void PrepareForUse()
+        {
+            base.PrepareForUse();
+            runAnimation();
+        }
+
+        protected override void FreeAfterUse()
+        {
+            base.FreeAfterUse();
+
+            // clean up transforms on free instead of on prepare as is usually the case
```

Comment:
Not a huge fan of this. Would rather have an `if (IsPresent)` check in `VisualiseSecondHit` blocking the application of the animation (and applying in `runAnimation` if required).

Reply:
Wasn't super sure where the presence check fit in all of this, but I have rewritten this logic in a slightly different way to avoid that cleanup being there (it resembles the DHO state update code a little now). I *think* it's a little better...?